packages/babel-plugin-proposal-decorators/src/transformer-2022-03.ts

```diff
@@ -1020,6 +1037,186 @@ function transformClass(
   return path;
 }
 
+function isProtoSetter(
+  node: t.Identifier | t.StringLiteral | t.BigIntLiteral | t.NumericLiteral,
+) {
+  return node.type === "Identifier"
+    ? node.name === "__proto__"
+    : node.value === "__proto__";
+}
+
+function isDecorated(node: t.Class | ClassDecoratableElement) {
+  return node.decorators && node.decorators.length > 0;
+}
+
+function shouldTransformElement(node: ClassElement) {
+  switch (node.type) {
+    case "ClassAccessorProperty":
+      return true;
+    case "ClassMethod":
+    case "ClassProperty":
+    case "ClassPrivateMethod":
+    case "ClassPrivateProperty":
+      return isDecorated(node);
+    default:
+      return false;
+  }
+}
+
+function shouldTransformClass(node: t.Class) {
+  return (
+    isDecorated(node) ||
+    node.body.body.some(node => shouldTransformElement(node))
+  );
+}
+
+// Todo: unify name references logic with helper-function-name
+function NamedEvaluationVisitoryFactory(
+  isAnonymous: (path: NodePath) => boolean,
+  visitor: (
+    path: NodePath,
+    state: PluginPass,
+    name: string | t.Identifier | t.StringLiteral,
+  ) => void,
+) {
+  return {
+    VariableDeclarator(path, state) {
+      const id = path.node.id;
+      if (id.type === "Identifier") {
+        const initializer = skipTransparentExprWrappers(path.get("init"));
+        if (isAnonymous(initializer)) {
+          const name = id.name;
+          visitor(initializer, state, name);
+        }
+      }
+    },
+    AssignmentExpression(path, state) {
+      const id = path.node.left;
+      if (id.type === "Identifier") {
+        const initializer = skipTransparentExprWrappers(path.get("right"));
+        if (isAnonymous(initializer)) {
+          switch (path.node.operator) {
+            case "=":
+            case "&&=":
+            case "||=":
+            case "??=":
+              visitor(initializer, state, id.name);
+          }
+        }
+      }
+    },
+    AssignmentPattern(path, state) {
+      const id = path.node.left;
+      if (id.type === "Identifier") {
+        const initializer = skipTransparentExprWrappers(path.get("right"));
+        if (isAnonymous(initializer)) {
+          const name = id.name;
+          visitor(initializer, state, name);
+        }
+      }
+    },
+    // We listen on ObjectExpression so that we don't have to visit
+    // the object properties under object patterns
+    ObjectExpression(path, state) {
```

Comment:
Guess I have to handle `RecordExpression` here as well.

Reply:
A record expression can not contain function as value anyway, so we can ignore that.

Reply:
@nicolo-ribaudo The current record spec requires NamedEvaluation for the initializer. https://tc39.es/proposal-record-tuple/#sec-record-initializer-runtime-semantics-recordpropertydefinitionevaluation Is that observable now that function / class can not be a record value?